### ISSUE TITLE
pdb inaccessible after inline collect

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [pytest-dev/pytest#1658]
just tried to use pdb after a inline collect and got the following

```
$ py.test testing/test_mark.py  -k test_mark_decorator_subclass_does_not_propagate_to_base2
=============================================================================================== test session starts ================================================================================================
platform linux2 -- Python 2.7.11, pytest-2.10.0.dev1, py-1.4.31, pluggy-0.3.1
rootdir: /home/ronny/Projects/pytest-dev/pytest, inifile: tox.ini
plugins: xdist-1.15.dev19+nge8fa737.d20160625, hypothesis-3.4.0
collected 59 items 

testing/test_mark.py F

===================================================================================================== FAILURES =====================================================================================================
_____________________________________________________________________ TestFunctional.test_mark_decorator_subclass_does_not_propagate_to_base2 ______________________________________________________________________

self = <test_mark.TestFunctional instance at 0x7f120eebfbd8>, testdir = <Testdir local('/tmp/pytest-of-ronny/pytest-184/testdir/test_mark_decorator_subclass_does_not_propagate_to_base20')>

    def test_mark_decorator_subclass_does_not_propagate_to_base2(self, testdir):
        p = testdir.makepyfile("""
                import pytest

                @pytest.mark.a
                class TestBase:
                    def test_foo(self):
                        pass

                @pytest.mark.b
                class Test2(TestBase):
                    pass

            """)
        items, rec = testdir.inline_genitems(p)
        print [(item, [x for x in item.keywords if item.get_marker(x) is not None]) for item in items]
>       pytest.set_trace()

/home/ronny/Projects/pytest-dev/pytest/testing/test_mark.py:438: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/ronny/Projects/pytest-dev/pytest/_pytest/pdb.py:44: in set_trace
    tw = _pytest.config.create_terminal_writer(self._config)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

config = None, args = (), kwargs = {}, tw = <py._io.terminalwriter.TerminalWriter object at 0x7f120edb6a90>

    def create_terminal_writer(config, *args, **kwargs):
        """Create a TerminalWriter instance configured according to the options
        in the config object. Every code which requires a TerminalWriter object
        and has access to a config object should use this function.
        """
        tw = py.io.TerminalWriter(*args, **kwargs)
>       if config.option.color == 'yes':
E       AttributeError: 'NoneType' object has no attribute 'option'

/home/ronny/Projects/pytest-dev/pytest/_pytest/config.py:1174: AttributeError
=============================================================== 58 tests deselected by '-ktest_mark_decorator_subclass_does_not_propagate_to_base2' ================================================================
===================================================================================== 1 failed, 58 deselected in 0.12 seconds ======================================================================================
```

when commenting out the inline-collect it wont complain

**Comment by fengxx** [comment]
I encounter same issue both in master and features branch

**Comment by RonnyPfannschmidt** [comment]
i wont be able to get deep enough into pytester so soon

**Comment by The-Compiler** [comment]
Is this a regression since the last release or an older bug? If it's an older thing, let's postpone it to get 3.0 out.

**Comment by RonnyPfannschmidt** [comment]
i suspect this is an older one, we already released a pytest where pytester has inline collect

**Comment by TristenHayfield** [comment]
I have been affected by this as well. It makes writing plugins a pain.

The issue is due to debugging.py:pytest_configure:fin() getting called after every round of inline tests.
fin() just wipes the old config away instead of restoring it.

changing pytest_configure to:

```python
def pytest_configure(config):
    p_old = pytestPDB()
    p_old._config = pytestPDB._config
    p_old._pdb_cls = pytestPDB._pdb_cls

    if config.getvalue("usepdb") or config.getvalue("usepdb_cls"):
        config.pluginmanager.register(PdbInvoke(), 'pdbinvoke')
        if config.getvalue("usepdb_cls"):
            modname, classname = config.getvalue("usepdb_cls").split(":")
            __import__(modname)
            pdb_cls = getattr(sys.modules[modname], classname)
        else:
            pdb_cls = pdb.Pdb
        pytestPDB._pdb_cls = pdb_cls

    old = (pdb.set_trace, pytestPDB._pluginmanager)

    def fin():
        pdb.set_trace, pytestPDB._pluginmanager = old
        pytestPDB._config = p_old._config
        pytestPDB._pdb_cls = p_old._pdb_cls

    pdb.set_trace = pytest.set_trace
    pytestPDB._pluginmanager = config.pluginmanager
    pytestPDB._config = config
    config._cleanup.append(fin)
```

fixes it.  It's a bit of a dirty fix - it would be better to just add methods to pytestPDB to clone and restore config from an instance.

I can do a PR and add some unit tests if you like.

**Comment by nicoddemus** [comment]
@TristenHayfield a PR would be welcome, we can discuss the finer details of the solution there.

**Comment by RonnyPfannschmidt** [comment]
#2329

**Comment by blueyed** [comment]
I assume this got fixed by e61e81a7, at least it mentions the same AttributeError.
